### PR TITLE
WebDav support

### PR DESCRIPTION
**Pull request by labkey-alan**
#### Rationale
This PR adds two methods to our ServerContext: webdav_client and webdav_path, which create a WebDav Client and construct a LabkeyServer WebDav path.

#### Related Pull Requests
* n/a

#### Changes
* Add `hostname` property to ServerContext
* Add `base_url` property to ServerContext
* Add `webdav_client` method to ServerContext
  * This method returns a webdavclient3 Client instance
* Add `webdav_path` method to ServerContext
* Add docs for WebDav support
* Add unit tests for ServerContext

### CODE REVIEW

**Comment by labkey-alan** [comment]
@labkey-nicka @cnathe I tested this locally with embedded tomcat, but have not tested using a server that has a context path. I am fairly certain I got this right but would be good to have someone test with a context path set on their LKS instance.

File: labkey/server_context.py [hunk 1]
```diff
@@ -81,23 +81,81 @@ def __init__(
...
+                path += "/"
+
+            path += file_name
+
+        return path
+
+    def webdav_client(self, webdav_options: dict = None):
+        # We localize the import of webdav3 here so it is an optional dependency. Only users who want to use webdav will
+        # need to pip install webdavclient3
+        from webdav3.client import Client
```
Comment (labkey-nicka) [inline labkey/server_context.py:136]: This is rad.

File: labkey/server_context.py [hunk 2]
```diff
@@ -81,23 +81,81 @@ def __init__(
...
+
+        if webdav_options is not None:
+            options = {
+                **options,
+                **webdav_options,
+            }
+
+        client = Client(options)
+
+        if self._verify_ssl is False:
```
Comment (labkey-nicka) [inline labkey/server_context.py:154]: Good catch.

**Review by labkey-nicka — APPROVED** [review]
Looks good and works well with my local testing. Here is the final script I utilized:

image

File: labkey/server_context.py (hunk 1 above)
Comment (labkey-alan) [reply labkey/server_context.py:136]: Yeah it's rare that his feature is useful, but it's powerful when used!

**Pull request merged**